server/src/main/java/org/elasticsearch/cluster/coordination/ClusterBootstrapService.java

```diff
@@ -21,214 +21,205 @@
     private final TimeValue unconfiguredBootstrapTimeout;
     private final TransportService transportService;
-    private volatile boolean running;
+    private final Supplier<Iterable<DiscoveryNode>> discoveredNodesSupplier;
+    private final BooleanSupplier isBootstrappedSupplier;
+    private final Consumer<VotingConfiguration> votingConfigurationConsumer;
+    private final AtomicBoolean bootstrappingPermitted = new AtomicBoolean(true);
+
+    public ClusterBootstrapService(Settings settings, TransportService transportService,
+                                   Supplier<Iterable<DiscoveryNode>> discoveredNodesSupplier, BooleanSupplier isBootstrappedSupplier,
+                                   Consumer<VotingConfiguration> votingConfigurationConsumer) {
+
+        final List<String> initialMasterNodes = INITIAL_MASTER_NODES_SETTING.get(settings);
+        bootstrapRequirements = new LinkedHashSet<>(initialMasterNodes);
+        if (bootstrapRequirements.size() != initialMasterNodes.size()) {
+            throw new IllegalArgumentException(
+                "setting [" + INITIAL_MASTER_NODES_SETTING.getKey() + "] contains duplicates: " + initialMasterNodes);
+        }
 
-    public ClusterBootstrapService(Settings settings, TransportService transportService) {
-        initialMasterNodes = INITIAL_MASTER_NODES_SETTING.get(settings);
         unconfiguredBootstrapTimeout = discoveryIsConfigured(settings) ? null : UNCONFIGURED_BOOTSTRAP_TIMEOUT_SETTING.get(settings);
         this.transportService = transportService;
+        this.discoveredNodesSupplier = discoveredNodesSupplier;
+        this.isBootstrappedSupplier = isBootstrappedSupplier;
+        this.votingConfigurationConsumer = votingConfigurationConsumer;
     }
 
     public static boolean discoveryIsConfigured(Settings settings) {
         return Stream.of(DISCOVERY_HOSTS_PROVIDER_SETTING, DISCOVERY_ZEN_PING_UNICAST_HOSTS_SETTING, INITIAL_MASTER_NODES_SETTING)
             .anyMatch(s -> s.exists(settings));
     }
 
-    public void start() {
-        assert running == false;
-        running = true;
+    void onFoundPeersUpdated() {
+        final Set<DiscoveryNode> nodes = getDiscoveredNodes();
+        if (transportService.getLocalNode().isMasterNode() && bootstrapRequirements.isEmpty() == false
+            && isBootstrappedSupplier.getAsBoolean() == false && nodes.stream().noneMatch(Coordinator::isZen1Node)) {
+
+            final Tuple<Set<DiscoveryNode>,List<String>> requirementMatchingResult;
+            try {
+                requirementMatchingResult = checkRequirements(nodes);
+            } catch (IllegalStateException e) {
+                logger.warn("bootstrapping cancelled", e);
+                bootstrappingPermitted.set(false);
+                return;
+            }
+
+            final Set<DiscoveryNode> nodesMatchingRequirements = requirementMatchingResult.v1();
+            final List<String> unsatisfiedRequirements = requirementMatchingResult.v2();
```

Comment:
can you output matching and unsatisfiedRequirements at trace level here?

Reply:
Yep, see 2878bd6